// @carto/airship-components/dist/collection/components/legends/as-legend-category-point-entry/as-legend-category-point-entry.js
import { borderStyleCounts } from '../../../utils/border-style-counts';
const FAKE_BORDER_SIZE = 1;
const DEFAULT_WIDTH = 16;
const SYMBOL_OFFSET = 2;
export class LegendCategoryPointEntry {
    constructor() {
        this.width = DEFAULT_WIDTH;
    }
    render() {
        const classes = {
            'as-legend-category-point-entry--point': !this.marker,
            'as-legend-category-point-entry--symbol': true
        };
        return (h("div", { class: 'as-legend-category-point-entry--wrapper' },
            h("div", { class: 'as-legend-category--figure--wrapper' },
                h("div", { class: classes, style: this.getStyle() })),
            h("span", { class: 'as-legend-category--label' }, this.label)));
    }
    getStyle() {
        return Object.assign({ backgroundColor: this.color, border: `${FAKE_BORDER_SIZE}px ${this.strokeStyle || 'solid'} ${this.strokeColor}` }, this.getWidth(), this.getMask());
    }
    getWidth() {
        if (!this.width) {
            return {};
        }
        let sizeOffset = borderStyleCounts(this.strokeStyle)
            ? FAKE_BORDER_SIZE * 2
            : 0;
        if (this.marker) {
            sizeOffset += SYMBOL_OFFSET;
        }
        return {
            height: `${this.width + sizeOffset}px`,
            width: `${this.width + sizeOffset}px`
        };
    }
    getMask() {
        if (!this.marker) {
            return {};
        }
        return {
            '-webkit-mask-image': `url(${this.marker})`,
            '-webkit-mask-position': 'center',
            '-webkit-mask-repeat': 'no-repeat',
            '-webkit-mask-size': `${this.width}px`,
            'maskImage': `url(${this.marker})`,
            'maskPosition': 'center',
            'maskRepeat': 'no-repeat',
            'maskSize': `${this.width}px`
        };
    }
    static get is() { return "as-legend-category-point-entry"; }
    static get properties() { return {
        "color": {
            "type": String,
            "attr": "color"
        },
        "label": {
            "type": String,
            "attr": "label"
        },
        "marker": {
            "type": String,
            "attr": "marker"
        },
        "strokeColor": {
            "type": String,
            "attr": "stroke-color"
        },
        "strokeStyle": {
            "type": String,
            "attr": "stroke-style"
        },
        "width": {
            "type": Number,
            "attr": "width"
        }
    }; }
    static get style() { return "/**style-placeholder:as-legend-category-point-entry:**/"; }
}
