// @carto/airship-components/dist/esm/es2017/build/kcyrxlpo.entry.js
import { h } from '../airship.core.js';

class LegendColorBins {
    render() {
        if (!this.data || this.data.length === 0) {
            return null;
        }
        if (this.data[0].type === 'polygon') {
            return (h("as-legend-color-bins-polygon", { data: this.data, orientation: this.orientation }));
        }
        else {
            return (h("as-legend-category", { data: this.data, orientation: this.orientation, width: this.width }));
        }
    }
    static get is() { return "as-legend-color-bins"; }
    static get properties() { return {
        "data": {
            "type": "Any",
            "attr": "data"
        },
        "orientation": {
            "type": String,
            "attr": "orientation"
        },
        "width": {
            "type": Number,
            "attr": "width"
        }
    }; }
    static get style() { return "as-legend-color-bins{display:block}"; }
}

class LegendColorBinsPolygon {
    constructor() {
        this.orientation = 'vertical';
    }
    render() {
        if (!this.data) {
            return null;
        }
        const outerClasses = {
            'as-legend-color-bins-polygon--outer-wrapper': true,
            [`as-legend-color-bins-polygon--${this.orientation}`]: true
        };
        return h("div", { class: outerClasses },
            h("div", { class: 'as-legend-color-bins-polygon--wrapper as-legend-color-bins-polygon--color' }, this.data.map(this.renderStep)),
            h("div", { class: 'as-legend-color-bins-polygon--wrapper as-legend-color-bins-polygon--labels' }, this.data.map(this.renderLabels)));
    }
    renderStep(data) {
        return (h("div", { class: 'as-legend-color-bins-polygon--step', style: { background: data.color } }));
    }
    renderLabels(data) {
        return (h("div", { class: 'as-legend-color-bins-polygon--label' },
            h("span", null, data.label)));
    }
    static get is() { return "as-legend-color-bins-polygon"; }
    static get properties() { return {
        "data": {
            "type": "Any",
            "attr": "data"
        },
        "orientation": {
            "type": String,
            "attr": "orientation"
        }
    }; }
    static get style() { return "as-legend-color-bins-polygon{--as-legend-color-bins-polygon--border-color:rgba(0,0,0,0.1);--as-legend-color-bins-polygon--color:var(--as--color--type-01);display:block}as-legend-color-bins-polygon .as-legend-color-bins-polygon--outer-wrapper{display:-ms-flexbox;display:flex}as-legend-color-bins-polygon .as-legend-color-bins-polygon--outer-wrapper.as-legend-color-bins-polygon--vertical{-ms-flex-direction:row;flex-direction:row}as-legend-color-bins-polygon .as-legend-color-bins-polygon--outer-wrapper.as-legend-color-bins-polygon--horizontal{-ms-flex-direction:column;flex-direction:column}as-legend-color-bins-polygon .as-legend-color-bins-polygon--wrapper{display:-ms-flexbox;display:flex}as-legend-color-bins-polygon .as-legend-color-bins-polygon--color{-webkit-box-shadow:-1px 0 2px 0 var(--as-legend-color-bins-polygon--border-color);box-shadow:-1px 0 2px 0 var(--as-legend-color-bins-polygon--border-color)}as-legend-color-bins-polygon .as-legend-color-bins-polygon--horizontal .as-legend-color-bins-polygon--wrapper{min-width:300px}as-legend-color-bins-polygon .as-legend-color-bins-polygon--horizontal .as-legend-color-bins-polygon--labels{margin-top:8px}as-legend-color-bins-polygon .as-legend-color-bins-polygon--label{-ms-flex:1;flex:1;color:var(--as-legend-color-bins-polygon--color);font:var(--as--font--caption);text-transform:capitalize;white-space:nowrap}as-legend-color-bins-polygon .as-legend-color-bins-polygon--vertical .as-legend-color-bins-polygon--wrapper{-ms-flex-direction:column;flex-direction:column;min-height:300px}as-legend-color-bins-polygon .as-legend-color-bins-polygon--vertical .as-legend-color-bins-polygon--label{display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column;-ms-flex-pack:center;justify-content:center}as-legend-color-bins-polygon .as-legend-color-bins-polygon--vertical .as-legend-color-bins-polygon--labels{margin-left:8px}as-legend-color-bins-polygon .as-legend-color-bins-polygon--step{-ms-flex:1;flex:1}as-legend-color-bins-polygon .as-legend-color-bins-polygon--horizontal .as-legend-color-bins-polygon--step{height:8px;border-top:1px solid var(--as-legend-color-bins-polygon--border-color);border-bottom:1px solid var(--as-legend-color-bins-polygon--border-color)}as-legend-color-bins-polygon .as-legend-color-bins-polygon--vertical .as-legend-color-bins-polygon--step{width:8px;border-right:1px solid var(--as-legend-color-bins-polygon--border-color);border-left:1px solid var(--as-legend-color-bins-polygon--border-color)}as-legend-color-bins-polygon .as-legend-color-bins-polygon--horizontal .as-legend-color-bins-polygon--step:first-of-type{border-left:1px solid var(--as-legend-color-bins-polygon--border-color)}as-legend-color-bins-polygon .as-legend-color-bins-polygon--horizontal .as-legend-color-bins-polygon--step:last-of-type{border-right:1px solid var(--as-legend-color-bins-polygon--border-color)}as-legend-color-bins-polygon .as-legend-color-bins-polygon--vertical .as-legend-color-bins-polygon--step:first-of-type{border-top:1px solid var(--as-legend-color-bins-polygon--border-color)}as-legend-color-bins-polygon .as-legend-color-bins-polygon--vertical .as-legend-color-bins-polygon--step:last-of-type{border-bottom:1px solid var(--as-legend-color-bins-polygon--border-color)}"; }
}

export { LegendColorBins as AsLegendColorBins, LegendColorBinsPolygon as AsLegendColorBinsPolygon };
