// @carto/airship-components/dist/esm/es2017/build/nycfox15.sc.entry.js
import { h } from '../airship.core.js';

import { a as borderStyleCounts } from './chunk-512e386e.js';

const FAKE_BORDER_SIZE = 1;
class LegendSizeBinsPoint {
    constructor() {
        this.orientation = 'vertical';
        this.width = null;
    }
    render() {
        if (!this.data) {
            return null;
        }
        const classes = {
            'as-legend-size-bins-point--steps': true,
            [`as-legend-size-bins-point--${this.orientation}`]: true
        };
        this.maxSize = this.width || this.data.slice().sort((first, second) => second.width - first.width)[0].width;
        return h("div", { class: classes }, this.data.map((data) => this.renderStep(data)));
    }
    renderStep(data) {
        const strokeStyle = `${FAKE_BORDER_SIZE}px ${data.strokeStyle || 'solid'} ${data.strokeColor}`;
        const sizeOffset = borderStyleCounts(data.strokeStyle)
            ? FAKE_BORDER_SIZE * 2
            : 0;
        const SIZE = Math.round(data.width) + sizeOffset;
        const SIZE_PX = `${SIZE}px`;
        const MAX_SIZE = SIZE > this.maxSize ? SIZE : this.maxSize + sizeOffset;
        const MAX_SIZE_PX = `${MAX_SIZE}px`;
        const mask = this.getMask(data);
        const wrapperStyle = {};
        if (this.orientation === 'horizontal') {
            wrapperStyle.height = MAX_SIZE_PX;
        }
        else if (this.orientation === 'vertical') {
            wrapperStyle.width = MAX_SIZE_PX;
        }
        const style = Object.assign({ backgroundColor: data.color, border: strokeStyle, height: SIZE_PX, width: SIZE_PX }, mask);
        return (h("div", { class: 'as-legend-size-bins-point--step' },
            h("div", { style: wrapperStyle, class: 'as-legend-size-bins-point--circle-wrapper' },
                h("div", { class: 'as-legend-size-bins-point--circle', style: style })),
            h("span", { class: 'as-legend-size-bins-point--label' }, data.label)));
    }
    getMask(data) {
        if (!data.marker) {
            return {};
        }
        return {
            '-webkit-mask-image': `url(${data.marker})`,
            '-webkit-mask-position': 'center',
            '-webkit-mask-repeat': 'no-repeat',
            '-webkit-mask-size': `${data.width}px`,
            'maskImage': `url(${data.marker})`,
            'maskPosition': 'center',
            'maskRepeat': 'no-repeat',
            'maskSize': `${data.width}px`
        };
    }
    static get is() { return "as-legend-size-bins-point"; }
    static get properties() { return {
        "data": {
            "type": "Any",
            "attr": "data"
        },
        "orientation": {
            "type": String,
            "attr": "orientation"
        },
        "width": {
            "type": Number,
            "attr": "width"
        }
    }; }
    static get style() { return "as-legend-size-bins-point{--as-legend-size-bins-point--color:var(--as--color--type-01);--as-legend-size-bins-point--shadow:rgba(0,0,0,0.1);display:block}as-legend-size-bins-point .as-legend-size-bins-point--steps{display:-ms-flexbox;display:flex}as-legend-size-bins-point .as-legend-size-bins-point--step{display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center}as-legend-size-bins-point .as-legend-size-bins-point--circle{border-radius:50%;-webkit-box-shadow:-1px 0 2px 0 var(--as-legend-size-bins-point--shadow);box-shadow:-1px 0 2px 0 var(--as-legend-size-bins-point--shadow)}as-legend-size-bins-point .as-legend-size-bins-point--circle-wrapper{display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;-ms-flex-pack:center;justify-content:center}as-legend-size-bins-point .as-legend-size-bins-point--label{display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;color:var(--as-legend-size-bins-point--color);font:var(--as--font--caption);text-transform:capitalize;white-space:nowrap}as-legend-size-bins-point .as-legend-size-bins-point--horizontal{-ms-flex-direction:row;flex-direction:row;-ms-flex-align:start;align-items:flex-start;-ms-flex-pack:distribute;justify-content:space-around}as-legend-size-bins-point .as-legend-size-bins-point--horizontal .as-legend-size-bins-point--step{-ms-flex-direction:column;flex-direction:column;margin:0 4px}as-legend-size-bins-point .as-legend-size-bins-point--horizontal .as-legend-size-bins-point--step:first-of-type{margin-left:0}as-legend-size-bins-point .as-legend-size-bins-point--horizontal .as-legend-size-bins-point--step:last-of-type{margin-right:0}as-legend-size-bins-point .as-legend-size-bins-point--horizontal .as-legend-size-bins-point--label{margin-top:8px}as-legend-size-bins-point .as-legend-size-bins-point--horizontal .as-legend-size-bins-point--circle-wrapper{-ms-flex-align:end;align-items:flex-end}as-legend-size-bins-point .as-legend-size-bins-point--vertical{-ms-flex-direction:column;flex-direction:column}as-legend-size-bins-point .as-legend-size-bins-point--vertical .as-legend-size-bins-point--label{margin-left:8px}as-legend-size-bins-point .as-legend-size-bins-point--vertical .as-legend-size-bins-point--step:not(:first-of-type){margin-top:8px}"; }
}

export { LegendSizeBinsPoint as AsLegendSizeBinsPoint };
