// @carto/airship-components/dist/collection/components/legends/as-legend-color-continuous-polygon/as-legend-color-continuous-polygon.js
export class LegendColorContinuousPolygon {
    constructor() {
        this.orientation = 'vertical';
    }
    render() {
        if (!this.data) {
            return null;
        }
        const outerClasses = {
            'as-legend-color-continuous-polygon--outer-wrapper': true,
            [`as-legend-color-continuous-polygon--${this.orientation}`]: true
        };
        return h("div", { class: outerClasses },
            h("div", { class: 'as-legend-color-continuous-polygon--wrapper as-legend-color-continuous-polygon--color' }, this.data.map((data, index, arr) => this.renderGradientStep(data, index, arr))),
            h("div", { class: 'as-legend-color-continuous-polygon--wrapper as-legend-color-continuous-polygon--labels' }, this.data.map((data, index, arr) => this.renderLabel(data, index, arr))));
    }
    renderGradientStep(data, index, arr) {
        if (index === arr.length - 1) {
            return null;
        }
        const start = data.color;
        const end = arr[index + 1].color;
        const direction = this.orientation === 'vertical' ? '.5turn' : '.25turn';
        const style = {
            backgroundImage: `linear-gradient(${direction}, ${start}, ${end})`
        };
        return (h("div", { class: 'as-legend-color-continuous-polygon--step', style: style }));
    }
    renderLabel(data, index, arr) {
        if (this.orientation === 'vertical' && index === 0) {
            return null;
        }
        if (this.orientation === 'horizontal' && index === arr.length - 1) {
            return null;
        }
        return (h("div", { class: 'as-legend-color-continuous-polygon--label' },
            h("span", null, data.label)));
    }
    static get is() { return "as-legend-color-continuous-polygon"; }
    static get properties() { return {
        "data": {
            "type": "Any",
            "attr": "data"
        },
        "orientation": {
            "type": String,
            "attr": "orientation"
        }
    }; }
    static get style() { return "/**style-placeholder:as-legend-color-continuous-polygon:**/"; }
}
