// @carto/airship-components/dist/collection/components/legends/as-legend-category-line-entry/as-legend-category-line-entry.js
const MAX_WIDTH = 16;
export class LegendCategoryLineEntry {
    constructor() {
        this.strokeStyle = 'solid';
        this.width = 1.5;
    }
    componentWillLoad() {
        this._widthChanged(this.width);
    }
    render() {
        return (h("div", { class: 'as-legend-category-line-entry--wrapper' },
            h("div", { class: 'as-legend-category--figure--wrapper' },
                h("div", { class: 'as-legend-category-line-entry--line', style: this.getStyle() })),
            h("span", { class: 'as-legend-category--label' }, this.label)));
    }
    _widthChanged(newValue) {
        this._width = Math.min(MAX_WIDTH, newValue);
    }
    getStyle() {
        return {
            borderTopColor: this.color,
            borderTopStyle: this.strokeStyle,
            borderTopWidth: `${this._width}`
        };
    }
    static get is() { return "as-legend-category-line-entry"; }
    static get properties() { return {
        "_width": {
            "state": true
        },
        "color": {
            "type": String,
            "attr": "color"
        },
        "label": {
            "type": String,
            "attr": "label"
        },
        "strokeStyle": {
            "type": String,
            "attr": "stroke-style"
        },
        "width": {
            "type": Number,
            "attr": "width",
            "watchCallbacks": ["_widthChanged"]
        }
    }; }
    static get style() { return "/**style-placeholder:as-legend-category-line-entry:**/"; }
}
