// @carto/airship-components/dist/collection/components/legends/as-legend-size-bins-line/as-legend-size-bins-line.js
export class LegendSizeBinsLine {
    constructor() {
        this.orientation = 'vertical';
        this.width = null;
    }
    render() {
        if (!this.data) {
            return null;
        }
        const outerClasses = {
            'as-legend-size-bins-line--outer-wrapper': true,
            [`as-legend-size-bins-line--${this.orientation}`]: true
        };
        const wrapperStyle = {
            width: this.width !== null && this.orientation === 'vertical' ? `${this.width}px` : null
        };
        return h("div", { class: outerClasses },
            h("div", { style: wrapperStyle, class: 'as-legend-size-bins-line--wrapper as-legend-size-bins-line--color' }, this.data.map((d) => this.renderStep(d))),
            h("div", { class: 'as-legend-size-bins-line--wrapper as-legend-size-bins-line--labels' }, this.data.map(this.renderLabels)));
    }
    renderStep(data) {
        const prop = this.orientation === 'vertical' ? 'width' : 'height';
        const style = {
            background: data.color,
            [prop]: `${data.width}px`
        };
        return (h("div", { class: 'as-legend-size-bins-line--step', style: style }));
    }
    renderLabels(data) {
        return (h("div", { class: 'as-legend-size-bins-line--label' },
            h("span", null, data.label)));
    }
    static get is() { return "as-legend-size-bins-line"; }
    static get properties() { return {
        "data": {
            "type": "Any",
            "attr": "data"
        },
        "orientation": {
            "type": String,
            "attr": "orientation"
        },
        "width": {
            "type": Number,
            "attr": "width"
        }
    }; }
    static get style() { return "/**style-placeholder:as-legend-size-bins-line:**/"; }
}
